Scatter Plot Panel

timio23 opened the pull request.

formfcw reviewed (changes requested):
Great work @timio23 🎉 

In addition to the comments in the code, here is a list of things I noticed while visually reviewing the extension. Please be sure to update the code to at least fix the bugs!

1. Decimal values do not affect the X-axis
    <details><summary>screenshot</summary>

    ![bug-decimal-values](https://github.com/user-attachments/assets/232af2ac-1a5c-412c-a3d2-b34f68f7ddc5)

    </details>

2. The dots do not align with the x-axis label when the field is a datetime or timestamp field
    <details><summary>screenshot</summary>

    ![datetime-int-dot-alignment](https://github.com/user-attachments/assets/9a089c3a-ef6e-4e17-9132-0af6a81253b2)

    </details>

3. Currently there are two tooltips, we could get rid of the bottom one
    <details><summary>screenshot</summary>

    ![double-tooltip](https://github.com/user-attachments/assets/f9ede07b-3988-498a-8519-9a1c5b8fc880)

    </details>

4. If you hover over a dot, it gets bigger and its opacity is reduced; if tooltips are disabled, it just reduces the opacity without getting bigger. Make sure you are consistent here. I would suggest just scaling the dot up a bit on hover or changing the opacity to 100%.
    <details><summary>screenshots</summary>

    ![hover-dot-tooltip](https://github.com/user-attachments/assets/199ce101-46a3-4c03-a4b9-2ed6cd94d820)
    ![hover-dot](https://github.com/user-attachments/assets/d738b8b8-e54f-45d4-93de-60c88266d273)

    </details>

5. When you click on a dot, it gets a gray border, which seems unintentional. You could set its opacity to 100% and remove the border (or make it white) instead.

6. You could change the format of the tooltips from …
    ```js
    `{{ valueOfXAxis }}`
    `{{ colorDot }} {{ collectionName }}: {{ valueOfYAxis }}`
    // or with Series Grouping enabled
    `{{ colorDot }} {{ seriesGroupName }}: {{ valueOfYAxis }}`
    ```
    … to …
    ```js
    `{{ fieldNameOfXAxis }}: {{ valueOfXAxis }}`
    `{{ colorDot }} {{ fieldNameOfYAxis }}: {{ valueOfYAxis }}`
    // or with Series Grouping enabled
    `{{ colorDot }} {{ seriesGroupName }} – {{ fieldNameOfYAxis }}: {{ valueOfYAxis }}`
    ```

timio23 commented:
Hi @BB-Loft, I’ve been overseas this week and just got home. I will look at this tonight

timio23 commented:
Thanks for your comments @formfcw, I've addressed most of the comments.

1. Wasn't able to reproduce this. X Axis working perfectly for me as integer, decimal and float. Could you provide more information on how you reached this result?
2. The X axis for datetime values are factoring the time when plotting. If I set the time points to exactly midnight, the dots line up perfectly. This should be more visible with the recent changes to the tooltips. (It seems the Directus datetime picker defaults to midday)
3. Bottom tooltip removed as requested
4. Dots now remain the same size. Opacity is the only change
5. Yes, this was unintentional. Stroke has now been removed when a dot is clicked.
6. Tooltip formatting was a little strict when I keep the automations in place. I went for a similar approach along the lines of what you suggested.

```
`{{ fieldNameOfXAxis }}: {{ valueOfXAxis }}`
`{{ colorDot }} {{ collectionName }}: {{ fieldNameOfYAxis }} {{ valueOfYAxis }}`
// Series Grouping enabled
`{{ colorDot }} {{ seriesGroupName }}: {{ fieldNameOfYAxis }} {{ valueOfYAxis }}`
```

Changes to dependencies, readme and index have been made as suggested.

formfcw commented on packages/scatter-plot-panel/package.json:
```diff
@@ -0,0 +1,36 @@
+		"link": "directus-extension link"
+	},
+	"devDependencies": {
+		"@directus/extensions-sdk": "9.23.3",
```
If there is no specific reason, please use the latest sdk version `12.1.0`.

formfcw commented on packages/scatter-plot-panel/src/panel.vue:
```diff
@@ -0,0 +1,353 @@
+						enabled: showToolTip,
+						x: {
+							formatter: (value: any) => {
+								let formattedValue = type == 'datetime' ? format(value, (xField.type == 'date' ? "MMM d" : "PPp")) : value > 10000
```
Make sure to be consistent with the date format: use `xField.type == 'date' ? "MMM d" : "MMM d, p"` or `xField.type == 'date' ? "PP" : "PPp"`

formfcw commented on packages/scatter-plot-panel/src/panel.vue:
```diff
@@ -0,0 +1,353 @@
+							formatter: (value: any) => {
+								let formattedValue = type == 'datetime' ? format(value, (xField.type == 'date' ? "MMM d" : "PPp")) : value > 10000
+									? abbreviateNumber(value, 1)
+									: n(value, 'decimal', {
+											minimumFractionDigits: 0,
+											maximumFractionDigits: 2,
+									} as any);
+								return `${xField.name}: ${formattedValue}`;
```
This `n` function causes an error when using decimals (for the x-axis). This error occurs when I hover over a dot. When I use type `float`, it works fine.

formfcw replied:
works with `sqlite3`

formfcw reviewed (changes requested):
Thanks @timio23! A few things …

~~> 1. Wasn't able to reproduce this. X Axis working perfectly for me as integer, decimal and float. Could you provide more information on how you reached this result?~~

~~I am still running into this problem. I'm using MySQL - maybe that's the reason. Can you test?~~

**(Update: see below)**

---

> 2. The X axis for datetime values are factoring the time when plotting. If I set the time points to exactly midnight, the dots line up perfectly. This should be more visible with the recent changes to the tooltips. (It seems the Directus datetime picker defaults to midday)

You are right with that the date picker defaults to midday, at the same time it still doesn’t align well, even if I set the value to 12 AM (`00:00`).

![](https://github.com/user-attachments/assets/70d7e18b-fbab-4c9d-8c1c-6d4a26419df0)

formfcw commented:
I have now tested with `sqlite3` where it works fine. So maybe this is just related to MySQL ...

formfcw commented:
OK, @timio23. Please make sure that decimal values (for x-axis) work at least with `sqlite3` and `Postgres` as db.

timio23 commented:
@formfcw, the datetime alignment seems to be a UTC issue. I found a parameter that disables UTC on the x-axis so it uses your devices local time instead. This should align to your chosen timepoints now. As for the decimals and floats, I've forced the value to a string in the hope it will standardize the data sent to Apex charts. Theoretically Knex should be returning the data in the same way regardless of the database.

formfcw reviewed (changes requested):
Thank you @timio23. I have tested your panel with postgres db and here I have the same problem with `decimal` values on the x-axis. All the dots align to zero and when I hover over a dot I get a console error. Please make sure it works.

timio23 commented:
@formfcw, I managed to replicate the issue. The api returns the floats and decimals as strings instead of numbers. Forced the strings to numbers by using * 1. Does this work for you as well?

formfcw reviewed (dismissed):
LGTM 🚀 Thank you!
